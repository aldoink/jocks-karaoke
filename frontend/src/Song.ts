export class Song {
    constructor(
        public location: string,
        public title: string,
        public artist: string) {}
}