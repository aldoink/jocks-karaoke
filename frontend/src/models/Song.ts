export class Song {
    constructor(
        public id: number,
        public location: string,
        public title: string,
        public artist: string
    ) {}
}