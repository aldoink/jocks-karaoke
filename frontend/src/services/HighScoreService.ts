import axios, {AxiosRequestConfig} from "axios";
import {AuthService, BACKEND_URL} from "./AuthService";

export interface HighScore {
    [index: string]: string | number | undefined;

    name?: string;
    score?: string;
    songId: number;
}

export class HighScoreService {
    constructor(private authService: AuthService) {
    }

    public async findAll(songId: number): Promise<HighScore[]> {
        const response = await axios.get(`${BACKEND_URL}/highscores/${songId}`, this.getAxiosConfig());
        return response?.data;
    }

    public async save(entry: HighScore) {
        return await axios.put(
            `${BACKEND_URL}/highscores`,
            entry,
            this.getAxiosConfig()
        )
    }

    private getAxiosConfig(): AxiosRequestConfig {
        return {headers: {Authorization: "Bearer " + this.authService.getToken()}};
    }
}