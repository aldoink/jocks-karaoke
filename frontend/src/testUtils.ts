import {act} from "@testing-library/react";

export async function flushPromises() {
    await act(async () => {
        await new Promise(resolve => setImmediate(resolve))
    })
}