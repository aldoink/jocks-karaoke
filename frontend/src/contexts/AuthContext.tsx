import {createContext} from "react";
import {AuthService} from "../services/AuthService";

export const AuthContext = createContext({authService: {} as AuthService});