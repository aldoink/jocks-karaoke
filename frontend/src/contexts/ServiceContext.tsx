import {createContext} from "react";
import {HighScoreService} from "../services/HighScoreService";
import {SongService} from "../services/SongService";

export interface IServiceContext {
    readonly highScoreService: HighScoreService
    readonly songService: SongService
}

export const ServiceContext = createContext<IServiceContext>(
    {
        highScoreService: {} as HighScoreService,
        songService: {} as SongService
    }
);

