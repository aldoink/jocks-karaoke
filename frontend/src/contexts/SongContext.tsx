import { createContext, Dispatch, SetStateAction } from "react";
import { SongService } from "../services/SongService";
import { Song } from "../models/Song";

export const songService = new SongService();
export interface ISongContext {
  readonly songService: SongService;
  songs: Song[];
  setSongs: Dispatch<SetStateAction<Song[]>>;
}

export const SongContext = createContext<ISongContext>({
  songService,
  songs: [],
  setSongs: () => {},
});
