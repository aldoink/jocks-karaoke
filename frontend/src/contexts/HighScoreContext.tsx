import { createContext } from "react";
import { HighScore, HighScoreService } from "../services/HighScoreService";
import { authService } from "./AuthContext";
import { Song } from "../models/Song";

export const highScoreService = new HighScoreService(authService);

export interface IHighScoreContext {
  highScoreService: HighScoreService;
  highScores: HighScore[];
  refreshHighScores: (song: Song) => void;
  hasError: boolean;
}

export const HighScoreContext = createContext<IHighScoreContext>({
  highScoreService,
  highScores: [],
  refreshHighScores: () => {},
  hasError: false,
});
