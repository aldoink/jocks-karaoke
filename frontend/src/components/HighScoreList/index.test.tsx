import {HighScoreList} from "./index";
import {Song} from "../../models/Song";
import {render, screen} from "@testing-library/react";
import {ServiceContext} from "../../contexts/ServiceContext";
import {HighScore, HighScoreService} from "../../services/HighScoreService";
import {waitFor} from "@testing-library/dom";
import React from "react";
import {flushPromises} from "../../testUtils";
import userEvent from "@testing-library/user-event";

describe('HighScoreList', () => {
    const defaultSong = new Song(1, "BP-0001", "Title", "Artist");
    let mockedHighScoreService: HighScoreService;

    function renderHighScoreList(song: Song = defaultSong) {
        return render(
            <ServiceContext.Provider value={{highScoreService: mockedHighScoreService}}>
                <HighScoreList song={song}></HighScoreList>
            </ServiceContext.Provider>
        )
    }

    beforeEach(() => {
        mockedHighScoreService = {} as HighScoreService;
    });

    it('loads the high scores for the selected song', async () => {
        //given
        mockedHighScoreService.findAll = jest.fn().mockResolvedValue([
            {name: 'Ally', score: 99, songId: 1} as HighScore,
            {name: 'Jenny', score: 98, songId: 1} as HighScore,
            {name: 'Johnny', score: 97, songId: 1} as HighScore,
            {name: 'Gemma', score: 96, songId: 1} as HighScore
        ]);
        const component = renderHighScoreList();

        //then
        await waitFor(() => expect(mockedHighScoreService.findAll).toHaveBeenCalledTimes(1))
        expect(mockedHighScoreService.findAll).toHaveBeenCalledWith(defaultSong.id);
        expect(component.asFragment()).toMatchSnapshot();
    });

    it("shows an error message when high scores couldn't be loaded", async () => {
        //given
        mockedHighScoreService.findAll = jest.fn().mockRejectedValue(new Error("Oops!"));
        renderHighScoreList();

        //then
        await flushPromises();
        expect(screen.getByText("Something went wrong... try again later.")).toBeInTheDocument();
    });

    it('shows a message when no high scores are found', () => {
        //given
        mockedHighScoreService.findAll = jest.fn().mockResolvedValue([]);
        renderHighScoreList();

        //then
        expect(screen.getByText("Doesn't look like anyone's set a high score yet!")).toBeInTheDocument();
    });
});