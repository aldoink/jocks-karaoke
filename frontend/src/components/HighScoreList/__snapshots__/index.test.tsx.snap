// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighScoreList loads the high scores for the selected song 1`] = `
<DocumentFragment>
  <div
    id="high-score-list"
  >
    <div
      class="title-container"
    >
      <h2>
        Title
      </h2>
      <h3>
        Artist
      </h3>
      <h4>
        BP-0001
      </h4>
    </div>
    <div>
      <h2>
        Doesn't look like anyone's set a high score yet!
      </h2>
      <button>
        Add high score
      </button>
    </div>
  </div>
</DocumentFragment>
`;
